x/wasm/internal/keeper/ioutil.go

```diff
@@ -0,0 +1,38 @@
+package keeper
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+)
+
+// magic bytes to identify gzip.
+// See https://www.ietf.org/rfc/rfc1952.txt
+// and https://github.com/golang/go/blob/master/src/net/http/sniff.go#L186
+var gzipIdent = []byte("\x1F\x8B\x08")
+
+// limit max bytes read to prevent gzip bombs
+const maxSize = 400 * 1024
+
+// uncompress returns gzip uncompressed content or given src when not gzip.
+func uncompress(src []byte) ([]byte, error) {
+	if len(src) < 3 {
+		return src, nil
+	}
+	in := io.LimitReader(bytes.NewReader(src), maxSize)
+	buf := make([]byte, 3)
+	if _, err := io.ReadAtLeast(in, buf, 3); err != nil {
+		return nil, err
+	}
+	if !bytes.Equal(gzipIdent, buf) {
+		return src, nil
+	}
+	zr, err := gzip.NewReader(io.MultiReader(bytes.NewReader(buf), in))
```

Comment:
You limit the input size with a Limit Reader, I think we need to wrap `zr` with a limit as well.
A zip bomb would have a relatively small input size (`src`) but a huge output size `ioutil.ReadAll(zr)`.

Reply:
ja, should have been limit on `zr` instead. 🐛